From ca3a9e31a560d630cf20286eb30d63ddafc0a05a Mon Sep 17 00:00:00 2001
From: Bence Balogh <bence.balogh@arm.com>
Date: Mon, 26 Feb 2024 14:47:25 +0100
Subject: [PATCH] Set RPC caller session SHM size for Corstone 1000 SMMGW

Set RPC caller session shared memory size so it fits the UEFI variable
index. Validate if SMM_GATEWAY_MAX_UEFI_VARIABLES * [descriptor size]
would fit into the shared memory size. Also align the heap size
accordingly.

Upstream-Status: Submitted [https://review.trustedfirmware.org/c/TS/trusted-services/+/27865]
Signed-off-by: Imre Kis <imre.kis@arm.com>
Signed-off-by: Bence Balogh <bence.balogh@arm.com>
---
 .../config/default-opteesp/CMakeLists.txt     | 32 +++++++++++++++----
 .../config/default-sp/CMakeLists.txt          | 31 ++++++++++++++----
 .../providers/arm/corstone1000/platform.cmake |  4 ++-
 3 files changed, 52 insertions(+), 15 deletions(-)

diff --git a/deployments/smm-gateway/config/default-opteesp/CMakeLists.txt b/deployments/smm-gateway/config/default-opteesp/CMakeLists.txt
index 7becb3999..897a8dabd 100644
--- a/deployments/smm-gateway/config/default-opteesp/CMakeLists.txt
+++ b/deployments/smm-gateway/config/default-opteesp/CMakeLists.txt
@@ -1,5 +1,5 @@
 #-------------------------------------------------------------------------------
-# Copyright (c) 2021-2023, Arm Limited and Contributors. All rights reserved.
+# Copyright (c) 2021-2024, Arm Limited and Contributors. All rights reserved.
 #
 # SPDX-License-Identifier: BSD-3-Clause
 #
@@ -24,7 +24,30 @@ set(SP_BIN_UUID_CANON "ed32d533-99e6-4209-9cc0-2d72cdd998a7")
 set(SP_FFA_UUID_CANON "${SP_BIN_UUID_CANON}")
 set(SP_BOOT_ORDER "8")
 
-set(SP_HEAP_SIZE "32 * 1024" CACHE STRING "SP heap size in bytes")
+#-------------------------------------------------------------------------------
+#  Set target platform to provide drivers needed by the deployment
+#
+#-------------------------------------------------------------------------------
+add_platform(TARGET "smm-gateway")
+
+# SMM variable and RPC caller settings
+set(SMM_GATEWAY_MAX_UEFI_VARIABLES 40 CACHE STRING "Maximum UEFI variable count")
+set(SMM_RPC_CALLER_SESSION_SHARED_MEMORY_SIZE 8192 CACHE STRING "RPC caller buffer size in SMMGW")
+
+# Validating settings
+# The UEFI variable index entry size is 168 bytes
+math(EXPR SHM_MIN "${SMM_GATEWAY_MAX_UEFI_VARIABLES} * 168")
+
+if (${SMM_RPC_CALLER_SESSION_SHARED_MEMORY_SIZE} LESS ${SHM_MIN})
+	message(FATAL_ERROR "The RPC SHM size must be at least 168 * [max UEFI variable count]")
+endif()
+
+target_compile_definitions("smm-gateway" PRIVATE
+	RPC_CALLER_SESSION_SHARED_MEMORY_SIZE=${SMM_RPC_CALLER_SESSION_SHARED_MEMORY_SIZE}
+	SMM_GATEWAY_MAX_UEFI_VARIABLES=${SMM_GATEWAY_MAX_UEFI_VARIABLES}
+)
+
+set(SP_HEAP_SIZE "16 * 1024 + ${SMM_GATEWAY_MAX_UEFI_VARIABLES} * 168 + ${SMM_RPC_CALLER_SESSION_SHARED_MEMORY_SIZE}" CACHE STRING "SP heap size in bytes")
 set(TRACE_PREFIX "SMMGW" CACHE STRING "Trace prefix")
 
 # Setting the MM communication buffer parameters
@@ -50,11 +73,6 @@ include(../../env/commonsp/smm_gateway_sp.cmake REQUIRED)
 include(../../infra/psa-varstore.cmake REQUIRED)
 include(../../smm-gateway.cmake REQUIRED)
 
-#-------------------------------------------------------------------------------
-#  Set target platform to provide drivers needed by the deployment
-#
-#-------------------------------------------------------------------------------
-add_platform(TARGET "smm-gateway")
 
 #-------------------------------------------------------------------------------
 #  Deployment specific build options
diff --git a/deployments/smm-gateway/config/default-sp/CMakeLists.txt b/deployments/smm-gateway/config/default-sp/CMakeLists.txt
index e56a8559d..d3a96b0c6 100644
--- a/deployments/smm-gateway/config/default-sp/CMakeLists.txt
+++ b/deployments/smm-gateway/config/default-sp/CMakeLists.txt
@@ -29,7 +29,30 @@ set(TRACE_PREFIX "SMMGW" CACHE STRING "Trace prefix")
 set(SP_STACK_SIZE "64 * 1024" CACHE STRING "Stack size")
 set(SP_BOOT_ORDER "8")
 
-set(SP_HEAP_SIZE "32 * 1024" CACHE STRING "Heap size")
+#-------------------------------------------------------------------------------
+#  Set target platform to provide drivers needed by the deployment
+#
+#-------------------------------------------------------------------------------
+add_platform(TARGET "smm-gateway")
+
+# SMM variable and RPC caller settings
+set(SMM_GATEWAY_MAX_UEFI_VARIABLES 40 CACHE STRING "Maximum UEFI variable count")
+set(SMM_RPC_CALLER_SESSION_SHARED_MEMORY_SIZE 8192 CACHE STRING "RPC caller buffer size in SMMGW")
+
+# Validating settings
+# The UEFI variable index entry size is 168 bytes
+math(EXPR SHM_MIN "${SMM_GATEWAY_MAX_UEFI_VARIABLES} * 168")
+
+if (${SMM_RPC_CALLER_SESSION_SHARED_MEMORY_SIZE} LESS ${SHM_MIN})
+	message(FATAL_ERROR "The RPC SHM size must be at least 168 * [max UEFI variable count]")
+endif()
+
+target_compile_definitions("smm-gateway" PRIVATE
+	RPC_CALLER_SESSION_SHARED_MEMORY_SIZE=${SMM_RPC_CALLER_SESSION_SHARED_MEMORY_SIZE}
+	SMM_GATEWAY_MAX_UEFI_VARIABLES=${SMM_GATEWAY_MAX_UEFI_VARIABLES}
+)
+
+set(SP_HEAP_SIZE "16 * 1024 + ${SMM_GATEWAY_MAX_UEFI_VARIABLES} * 168 + ${SMM_RPC_CALLER_SESSION_SHARED_MEMORY_SIZE}" CACHE STRING "SP heap size in bytes")
 
 # Setting the MM communication buffer parameters
 set(MM_COMM_BUFFER_ADDRESS "0x00000008 0x81000000" CACHE STRING "Address of MM communicte buffer in 64 bit DTS format")
@@ -49,12 +72,6 @@ include(../../env/commonsp/smm_gateway_sp.cmake REQUIRED)
 include(../../infra/psa-varstore.cmake REQUIRED)
 include(../../smm-gateway.cmake REQUIRED)
 
-#-------------------------------------------------------------------------------
-#  Set target platform to provide drivers needed by the deployment
-#
-#-------------------------------------------------------------------------------
-add_platform(TARGET "smm-gateway")
-
 #-------------------------------------------------------------------------------
 #  Deployment specific build options
 #-------------------------------------------------------------------------------
diff --git a/platform/providers/arm/corstone1000/platform.cmake b/platform/providers/arm/corstone1000/platform.cmake
index d16cde3f4..fd93d6f7e 100644
--- a/platform/providers/arm/corstone1000/platform.cmake
+++ b/platform/providers/arm/corstone1000/platform.cmake
@@ -9,9 +9,11 @@
 # include MHU driver
 include(${TS_ROOT}/platform/drivers/arm/mhu_driver/component.cmake)
 
+set(SMM_GATEWAY_MAX_UEFI_VARIABLES 80 CACHE STRING "Maximum UEFI variable count")
+set(SMM_RPC_CALLER_SESSION_SHARED_MEMORY_SIZE 16384 CACHE STRING "RPC caller buffer size in SMMGW")
+
 target_compile_definitions(${TGT} PRIVATE
 	SMM_VARIABLE_INDEX_STORAGE_UID=0x787
-	SMM_GATEWAY_MAX_UEFI_VARIABLES=80
 )
 
 add_compile_definitions(MBEDTLS_ECP_DP_SECP521R1_ENABLED)
-- 
2.25.1


